core/group: add allowedBlocks attribute to latest deprecation

sque89 opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Fix allowedBlocks attribute of the core/group block

## Why?
When adding a block to a post where the template includes a core/group block with the allowedBlocks attribute, the restriction works just fine the same instance and only the allowed blocks are able to be added as a child.

But once a new editor instance is created (reload the editor), the allowedBlocks attribute is completely ignored.

The reason is, that the `migrate` function of the latest deprecation does not receive the allowedBlocks attribute if it is not defined within its own attributes definition.

## How?
I am not not sure if this commit is the correct solution to the problem. It fixes it, but it seems weird to me that i need to add a new attribute to a deprecated version in order to make it persitent during the migration.

Could someone please have a look and tell if that is expected behaviour and therefore the solution to the issue?

## Testing Instructions

1. create a custom block which has a template with a core/group block in it which makes use of the allowedBlocks attribute and which sets the layout attribute of the group like follows: ```"layout":{"inherit":true}}```
2. add that block to a post and check that only the blockTypes in the allowedBlocks are available
3. save the post and reload the editor
4. check again which blockTypes can be added as a child to the group. Now all available blocks are available and the allowedBlocks attribute has no effect anymore

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->
![image](https://github.com/WordPress/gutenberg/assets/5357139/10c82a3e-f6a1-4991-ab07-23323019c746)

Commit:
core/group: add allowedBlocks attribute to latest deprecation

talldan commented:
I personally couldn't reproduce the issue when following the steps. It's working as expected for me.

The addition of `allowedBlocks` to the group block postdates this particular deprecation, so I wouldn't expect it to need to be added.

@sque89 You may need to narrow down the steps for reproduction. Are you using a clean install of WordPress/Gutenberg when testing or do you have other plugins installed?

sque89 commented:
@talldan thanks a lot for having a look and you are right, i missed a crucial precondition to this issue, sorry. I updated the steps to reproduce and added a screenshot of the relevant part in the code (it is the deprecated.js of the core/group block.

My custom blocks still use the layout inherit option which was apparently deprecated which is the reason why the latest deprecation is eligible for my group blocks.

If that happens, in the current version, the attributes passed into this deprecation migration does not contain the allowedBlocks attribute, so it gets omitted completely.

Adding the attrigute to this deprecation fixes the issue, but as said i am not sure if that is the right way to fix it, or if there is maybe a broader underlaying issue of attribute migration.

sque89 commented:
@talldan would appreciate if you may could have a look at my answer to your comment. This one is a pretty big problem for us currently. Thanks!

talldan commented:
@sque89 Sorry for the delay responding, it's been a busy time.

It would definitely be great to get this issue fixed, though I'm unsure what the right way to fix it is.

My understanding is a deprecation should represent a version of a block almost as a snapshot in time, so if the block didn't have `allowedBlocks` when the deprecation was created, it shouldn't need it.

I guess what's happening here is that the `isElligible` method is still causing the deprecation to run even on newer versions of the block 🤔 

Deprecations are a generally tricky area—I'll ask for some other input to see if we can help this one across the line.

tellthemachines commented:
Hello there @sque89,

`"layout":{"inherit":true}` was deprecated while layout was still an experimental feature, so there's no back compat requirement for it. Even so, back compat has been maintained for it to avoid breakage of existing websites, and that's why it still works for your custom blocks.

The stable equivalent layout is `layout: { type: 'constrained' }`. If you update your custom blocks to use that format instead of `"layout":{"inherit":true}`, everything should work as expected.

sque89 commented:
@talldan thanks for getting back on this. But i need to admit that i never looked at it the way @tellthemachines described it. If the feature never left the experimental state and was deprecated in that state, of course this needs to be considered as bad luck. That is exactly the risk when using experimental features.

So after considering that it would be fine for me to ignore this and i will handle the issue myself on our side by updating the blocks.

Thanks to you too for your time, i will close this PR.

talldan commented:
Thanks @sque89, glad a solution was found for you.
